### ISSUE TITLE
Durable Function fan-out (with Activity Triggers) scales too slowly

### ISSUE THREAD

**Issue by scott-aptuity** [Azure/azure-functions-durable-extension#1073]
Hi,
We are prototyping the standard Durable function Fan-Out scenario.
We have a loop that calls around 700 Activity Functions and then waits for them to complete.

My issue is that the activity functions don't appear to causing the functions to scale out - at least not any any reasonable amount of time.  

I see HTTP triggers will scale every 1 second, but other are 30 seconds?  
My Durable function is HTTP triggered and a I want to do a lot of identical work in parallel.  If the Fan-Out situation doesn't cause the function to scale, I'm not sure how useful it is.

Am I missing something or is Fan-Out not good at scaling out?

Thanks,
Scott

**Comment by ChuckEdx** [comment]
@cgillum an example would be instance id dd81b56c5b6d4084904a6cf5b548442b, Central US region

**Comment by olitomlinson** [comment]
> The reasons we originally had that disparity are no longer valid so we're looking into fixing soon so that non-HTTP workloads can scale as responsively as the HTTP workloads.

@cgillum when is this likely to land? this would be incredible for my use-case!

**Comment by mikestichling** [comment]
Hi,
We have noticed a similar issue to @ChuckEdx while prototyping Durable Functions for our use case. What we are seeing is that queueing up 100s or 1000s of activities from an orchestrator produces unexpected processing times. Below is a graph of 10 independent runs where an orchestrator was initiated from a HTTP trigger. The orchestrator calls an activity 1000 times asynchronously and waits for each activity to complete ([the code sample for](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-overview?tabs=csharp) the Fan-Out scenario was only slightly modified to track execution/queue times). The time an activity is called is passed to each activity as its input. All the activity does is record the time it was able to process and send back the stats as output. Nothing intensive is being done. Each activity on average takes a few ticks to complete.

<img width="1031" alt="Screen Shot 2019-12-18 at 11 36 40 am" src="https://user-images.githubusercontent.com/1830613/71046116-5d6fcc00-218b-11ea-9f75-79764257210b.png">

The y-axis is duration in seconds the activity waited to be processed. The x-axis is the activity index. In the runs above some activities waited a maximum of 30 seconds before they were processed. I have included a snippet of the code used to generate the above data for the graph:

```
public static class DurableFunctionBenchmark
    {
        [FunctionName("Orchestrator")]
        public static async Task<List<ActivityOutput>> RunOrchestrator(
            [OrchestrationTrigger] IDurableOrchestrationContext context)
        {
            var outputs = new List<ActivityOutput>();
            var parallelTasks = new List<Task<ActivityOutput>>();

            var contextInput = context.GetInput<FunctionInput>();
            var amountToProcess = contextInput.Amount;
            var enableParallelism = contextInput.EnableParallelism;
            var degreeOfParallelism = contextInput.DegreeOfParallelism;

            for (int i = 0; i < amountToProcess; i++)
            {
                if (enableParallelism && parallelTasks.Count(p => !p.IsCompleted) >= degreeOfParallelism)
                {
                    await Task.WhenAny(parallelTasks.Where(p => !p.IsCompleted));
                }

                Task<ActivityOutput> task = context.CallActivityAsync<ActivityOutput>("FanOut", new ActivityInput { ExecutedAt = DateTime.Now.ToString(), Index = i + 1, TestCase = contextInput.TestCase });
                parallelTasks.Add(task);
            }

            await Task.WhenAll(parallelTasks);

            outputs.AddRange(parallelTasks.Select(t => t.Result));
            return outputs;
        }

        [FunctionName("FanOut")]
        public static async Task<ActivityOutput> FanOut(
                [ActivityTrigger] ActivityInput input, ILogger log,
                [DurableClient] IDurableEntityClient client)
        {
            var start = DateTime.Now;

            var end = DateTime.Now;
            return new ActivityOutput(input.Index, DateTime.Parse(input.ExecutedAt), start, end, input.TestCase);
        }

        [FunctionName("FunctionTest_HttpStart")]
        public static async Task<HttpResponseMessage> HttpStart(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")]HttpRequestMessage req,
            [DurableClient] IDurableOrchestrationClient client,
            ILogger log)
        {
            var content = JsonConvert.DeserializeObject<FunctionInput>(await req.Content.ReadAsStringAsync());
            // Function input comes from the request content.
            string instanceId = await client.StartNewAsync("Orchestrator", null, content);
            log.LogInformation($"Started orchestration with ID = '{instanceId}'.");

            return client.CreateCheckStatusResponse(req, instanceId);
        }
    }
```

It might also be important to note that introducing a throttle on the number of activities being enqueued at a time didn’t significantly reduce the total amount of time taken to process all activities. The wait times in the queues were reduced. This can be seen in the graph below were 3 runs: 

<img width="906" alt="Screen Shot 2019-12-18 at 12 05 35 pm" src="https://user-images.githubusercontent.com/1830613/71047158-d1f83a00-218e-11ea-9e14-2cb23f797f86.png">

What isnt visible above is the time taken while waiting for activities to dequeue and complete. 

- Run1 with 10 batches at a time
- Run2 with no batching at all. Enqueueing all activities as quickly as possible.
- Run3 with 100 batches at a time

We would ideally like to kick off 1000s maybe even 10000s of activites from an orchestrator to then aggregate the results once all activities are complete. I cannot find any documentation on what the best practice is when calling activities from an orchestrator.

AppID: 5770f21b-6583-465c-90b8-0ecd8b6b0038
Region: Australia Southeast
Times: 2019-12-11 -> 2019-12-18. Data for the graphs were generated on the 2019-12-18 AEDT

**Comment by ChuckEdx** [comment]
@cgillum Is there any update on this?

**Comment by cgillum** [comment]
@ChuckEdx There are two different issues being discussed on this thread and unfortunately your specific case fell off my radar. Would you mind opening an Azure support ticket for your case, and provide a more recent instance where you observed that your app wasn't scaling?  Alternatively if you open a new GitHub issue with a recent repro we can try looking into it there, but a support ticket will provider a higher guarantee of a resolution.

**Comment by ChuckEdx** [comment]
@cgillum No worries. I meant in terms of the activity trigger scaling, because that's relevant for me as well.

**Comment by cgillum** [comment]
Ah, you were asking about the speed of scale? We're still having internal discussions about this. The main concern is making sure rapid scale out for non-HTTP workloads doesn't create too much capacity pressure. This is something we're currently planning on looking at over the next 6 months or so.

In the meantime, there are a couple workarounds you can consider:

* Use an **App Service Plan** with a fixed set of VM instances and/or include scheduled based auto-scale rules. Unfortunately this is not very "serverless" but it is an option that gives you more control over the resources your app uses.
* Use an **Elastic Premium Plan** and configure a minimum instance count. The premium VMs are more powerful and can have as many as 4 cores. They also scale a bit faster than the consumption plan, so you effectively get a much faster rate of CPU scaling with a configurable floor.

These options are not cheap, but they are options. We will definitely make public mention of when the faster scaling becomes available for the Consumption plan.

**Comment by ChuckEdx** [comment]
@cgillum Yes. Thank you for the update. I've tried the App Service Plan route, and I can confirm that it does scale rapidly. However, as you said, it's not very "serverless" and is probably less cost efficient. I look forward to seeing what you guys come up with for the Consumption Plan.

**Comment by emanuel-virca** [comment]
Any updates on this?

**Comment by lilyjma** [comment]
@cgillum - Chris do we have any updates that we can give here?

**Comment by cgillum** [comment]
The Azure Functions team is currently investing in some major updates to the scale infrastructure of the hosted service. These investments are expected to improve overall capability for all function types, including Durable Functions. I expect there will be some public announcements about this sometime later this year.

**Comment by YunchuWang** [comment]
@cgillum, do you have any updates from the Azure Functions team on standard Consumption non-HTTP scale-out cadence?

The current documentation still says that, for non-HTTP triggers, the platform allocates new instances at most once every 30 seconds: https://learn.microsoft.com/azure/azure-functions/event-driven-scaling#understanding-scaling-behaviors

The extension now has target-based scaling, and Flex Consumption uses a shorter worker-side queue polling interval, but neither appears to change the ordinary Consumption platform allocation cadence. Can you confirm whether the scale-infrastructure improvements mentioned in 2023 shipped for this scenario, and whether this issue can now be closed or should remain as a Functions platform tracking issue?

**Comment by cgillum** [comment]
No changes are planned for the Legacy (Windows) Consumption plans, nor for Elastic Premium plans at this time. If possible, customers should migrate to the newer Flex Consumption plan (the major investment I was hinting at in my earlier response), which supports target-based burst scaling. The scale improvements should be quite significant!
